### ISSUE TITLE
Support OIDC authentication

### ISSUE THREAD

**Issue by arbielsk** [microsoft/mindaro#80]
**Is your feature request related to a problem? Please describe.** 
We are currently evaluating using B2K for several teams in our company. One of these teams is working with a Kubernetes cluster that uses an OIDC authprovider. Unfortunately it seems like the C# kubernetes client that is used by B2K does not support OIDC.

The error that we receive is "User: {userDetails.Name} does not have appropriate auth credentials in kubeconfig" which seems to originate here: https://github.com/kubernetes-client/csharp/blob/master/src/KubernetesClient/KubernetesClientConfiguration.ConfigFile.cs 
 
**Describe the solution you'd like**
Ideally I'd like OIDC auth provider to be supported by Bridge To Kubernetes. More specifically I'd be interested in understanding if you have any plans for this at all at the current time. If not, would you consider supporting it, if it was supported by the C# client?
 
**Describe alternatives you've considered**
I don't think this is applicable.
 
**Additional context**
I'd be happy to provide more context including logs if that helps. Since these might contain PII I'd prefer to send them on a different channel than github.

Thank you for your time and consideration!

**Comment by greenie-msft** [comment]
Hi @arbielsk,

Thank you for your request and the details you've provided. 

I've created a work item to discuss your request with the larger team. It would also be helpful if you could send your logs to our email alias: [BridgeToKubernetes@microsoft.com](BridgeToKubernetes@microsoft.com). The information from the logs will help us identity the work needed to support this request.

We will respond to this thread when we have an update to share. Thank you.

**Comment by daniv-msft** [comment]
@arbielsk Thanks for reporting this issue. Could you please let us know which OIDC authprovider this other team in your company uses?
We are aware of a very similar issue on AKS clusters with AAD managed authentication enabled (cf. this issue: https://github.com/microsoft/mindaro/issues/26).

Also, does running standard `kubectl`commands work in their case? If yes, then it might be a similar problem as the one mentioned in this other issue, where we have to authenticate the cluster connection once (through `kubectl`) on the kubeconfig we target before being able to use it.
Here are the instructions for Visual Studio:
```
1. Ensure you have kubectl installed (either run az aks install-cli if you have the azure cli installed or follow the directions at https://kubernetes.io/docs/tasks/tools/install-kubectl/)
2. Find the cached kubeconfig the Bridge to Kubernetes VS extension is using to connect to your cluster: %LOCALAPPDATA%\BridgeToKubernetes\cache<subscription_id><resourcegroup_name><cluster_name>.kubeconfig
3. Open a new console or powershell window
4. Set a temporary KUBECONFIG environment variable in your terminal pointing to the the kubeconfig file from step 2 (set KUBECONFIG="<kubeconfig_path>" for cmd or $env:KUBECONFIG="<kubeconfig_path>" for powershell)
5. Run a kubectl command from the terminal session such as kubectl get namespaces
6. Follow the instructions from kubectl for authenticating the cluster connection
7. Once the authenticaiton process is complete, return to VS and refresh the cluster namespaces (or relaunch the configuration dialog if you closed it) and you should be able to access the cluster resources
```

For VS Code, you can just run `kubectl get namespaces` on the current kubeconfig targeted in the Kubernetes extension (in most of the cases, it should just be your default kubeconfig in `%UserProfile%\.kube\config`).

**Comment by arbielsk** [comment]
@greenie-msft Thank you for your response. I will provide more details via mail as soon as I've been able to test @daniv-msft's solution.

@daniv-msft Thank you for your reply! I'd love to test the workaround, however I am unsure where I can find the equivalent of %LOCALAPPDATA%\BridgeToKubernetes\cache<subscription_id><resourcegroup_name><cluster_name>.kubeconfig on MacOS. Could you please share that information with me so I can give it a shot? Thanks in advance!

**Comment by daniv-msft** [comment]
@arbielsk Thanks for your reply! If they are working on MacOS, could you please confirm that they are using VS Code and not Visual Studio?
If they are indeed using VS Code, the instructions are simpler: just making sure that the current kubeconfig set in the Kubernetes extension in VS Code is the right one, and running a kubect command on this kubeconfig (`kubectl get namespaces`).
If the problem is the same as mentioned in the other issue, then kubectl should trigger a one-time authentication flow.

**Comment by arbielsk** [comment]
@daniv-msft I have been able to confirm with my team that they are:

- [x] using VS Code (and not Visual Studio)
- [x] current kubeconfig set in the Kubernetes extension in VS Code is the right one. We checked this by explicitly setting the kubeconfig to **~/.kube/config**. The clusters/contexts shown were the correct ones and also the namespace was selected correctly.
- [x] run `kubectl get namespaces`. This worked correctly but did not result in a one-time authentication flow as the credentials were already available in the **~/.kube/config**

The problem stil persists.

I therefore expect that this is a different problem, related to the C# kubernetes client - I think this becomes apparent from the logs that I will send you in a minute via email referencing this github issue.

**Comment by daniv-msft** [comment]
Thanks @arbielsk! I replied to your email.

**Comment by creydr** [comment]
Hi,
could you please share your outcomes/status here as well, as I am interested in the OIDC integration, too.

Thanks

**Comment by daniv-msft** [comment]
@creydr Thanks for your message. As of today, we have a dependency on the [C# KubernetesClient](https://github.com/kubernetes-client/csharp/). I created this GitHub issue on their side, and unfortunately OIDC authproviders aren't supported on their side yet: https://github.com/kubernetes-client/csharp/issues/524

We're investigating other possibilities to get rid of this issue, but because of that as of today Bridge won't work with OIDC authproviders. In order to help us prioritize finding a workaround, could you please let me know if not having Bridge to Kubernetes working with OIDC is a blocker for you and your team?

**Comment by creydr** [comment]
Hi @daniv-msft,
thanks for your reply. We have a similar setup as you described in kubernetes-client/csharp#524. 
I tried to resolve it to use `kubectl proxy`, configure a context for this proxy and use this for the setup. But in this case I get a BadRequest:

From bridge-library-x.txt
```
...
2020-11-24T14:25:41.0100347Z | Library | ERROR | Logging handled exception: Microsoft.Rest.HttpOperationException: {"Request":{"Method":{"Method":"POST"},"RequestUri":"http://127.0.0.1:8001/api/v1/namespaces/my-namespace/serviceaccounts","Properties":{},"Content":"{\n  \"apiVersion\": \"v1\",\n  \"kind\": \"ServiceAccount\",\n  \"metadata\": {\n    \"labels\": {\n      \"mindaro.io/version\": \"v2\",\n      \"mindaro.io/component\": \"lpkrestorationjob\"\n    },\n    \"name\": \"lpkrestorationjob-v2\",\n    \"namespace\": \"my-namespace\"\n  }\n}","Headers":{"Content-Type":["application/json; charset=utf-8"],"Content-Length":["246"]}},"Response":{"StatusCode":400,"ReasonPhrase":"Bad Request","Content":"{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"the object provided is unrecognized (must be of type ServiceAccount): couldn't get version/kind; json parse error: unexpected end of JSON input (5b0a20207b0a202020202276616c7565223a205b5d2c0a2020202022706 ...)\",\"reason\":\"BadRequest\",\"code\":400}\n","Headers":{"Server":["SimpleHTTP/0.6","Python/3.6.9"],"Date":["Tue, 24 Nov 2020 14:25:40 GMT","Tue, 24 Nov 2020 14:25:40 GMT"],"Audit-Id":["fc19d11d-a5cd-4092-b7eb-502e3fffb8b0"],"Cache-Control":["no-cache, private"],"Content-Length":["325"],"Content-Type":["application/json"]}},"Body":null,"StackTrace":"   at k8s.Kubernetes.CreateNamespacedServiceAccountWithHttpMessagesAsync(V1ServiceAccount body, String namespaceParameter, String dryRun, String fieldManager, String pretty, Dictionary`2 customHeaders, CancellationToken cancellationToken)\n   at k8s.KubernetesExtensions.CreateNamespacedServiceAccountAsync(IKubernetes operations, V1ServiceAccount body, String namespaceParameter, String dryRun, String fieldManager, String pretty, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Common.Kubernetes.KubernetesClient.<>c__DisplayClass37_0.<<CreateServiceAccountIfNotExists>b__0>d.MoveNext()","Message":"Operation returned an invalid status code 'BadRequest'","Data":{},"InnerException":null,"HelpLink":null,"Source":"KubernetesClient","HResult":-2146233088}
2020-11-24T14:25:41.0179341Z | Library | ERROR | CreateServiceAccountIfNotExists threw HttpOperationException: StatusCode='BadRequest', ReasonPhrase='Bad Request', Content='{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the object provided is unrecognized (must be of type ServiceAccount): couldn't get version/kind; json parse error: unexpected end of JSON input (5b0a20207b0a202020202276616c7565223a205b5d2c0a2020202022706 ...)","reason":"BadRequest","code":400}\n'
2020-11-24T14:25:41.0189061Z | Library | TRACE | Event: RemoteRestoreJobDeployer-EnsureRbacResources <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":554.0}}</json>
2020-11-24T14:25:41.0197006Z | Library | TRACE | Event: RemoteRestoreJobDeployer-Deploy <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":558.0}}</json>
2020-11-24T14:25:41.2942641Z | Library | TRACE | Dependency: Kubernetes <json>{"target":"PatchV1DeploymentAsync","success":true,"duration":null,"properties":{}}</json>
2020-11-24T14:25:41.3049017Z | Library | TRACE | Event: WorkloadRestorationService-RestoreDeployment <json>{"properties":{"result":"Succeeded"},"metrics":{"durationInMs":279.0}}</json>
2020-11-24T14:25:41.3067041Z | Library | ERROR | Failed to deploy remote restore job for deployment my-namespace/ticket-service. Operation returned an invalid status code 'BadRequest'
2020-11-24T14:25:41.3095422Z | Library | TRACE | Event: KubernetesRemoteEnvironmentManager-PatchDeployment <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":2356.0}}</json>
2020-11-24T14:25:41.3108054Z | Library | TRACE | Event: KubernetesRemoteEnvironmentManager-StartRemoteAgent <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":2365.0}}</json>
2020-11-24T14:25:41.3111429Z | Library | TRACE | Event: ConnectManagementClient-StartRemoteAgent <json>{"properties":{"agentHostingMode":"2","sourceEntityType":"1","namespaceName":"my-namespace","containerName":"ticket-service","serviceName":"ticket-service","deploymentName":"ticket-service","podName":"ticket-service-7b7754b7dc-cf8nz","routingHeaderValue":"","localProcessConfigFilePath":"KubernetesLocalProcessConfig.yaml","result":"Failed"},"metrics":{"durationInMs":4183.0}}</json>
2020-11-24T14:25:41.3118009Z | Library | ERROR | Failed to start remote agent.
2020-11-24T14:25:41.3356015Z | Library | ERROR | Logging handled exception: Microsoft.DevSpaces.Common.Exceptions.UserVisibleException: {"Request":null,"Response":null,"RequestId":null,"ClientRequestId":null,"CorrelationRequestId":null,"Format":"Failed to deploy remote restore job for pod deployment 'my-namespace/ticket-service' with error: Operation returned an invalid status code 'BadRequest'","Args":[],"StackTrace":"   at Microsoft.DevSpaces.Library.Connect.KubernetesRemoteEnvironmentManager._PatchDeploymentAsync(RemoteContainerConnectionDetails remoteContainerConnectionDetails, IProgress`1 progress, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Library.Connect.KubernetesRemoteEnvironmentManager.StartRemoteAgentAsync(RemoteContainerConnectionDetails remoteContainerConnectionDetails, ILocalProcessConfig localProcessConfig, IProgress`1 progress, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Library.ManagementClients.ConnectManagementClient.<>c__DisplayClass18_0.<<StartRemoteAgentAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.DevSpaces.Library.ManagementClients.DevSpacesManagementClientExceptionStrategy.RunWithHandlingAsync[T](Func`1 func, FailureConfig failureConfig)","Message":"Failed to deploy remote restore job for pod deployment 'my-namespace/ticket-service' with error: Operation returned an invalid status code 'BadRequest'","Data":{},"InnerException":{"Request":{"Method":{"Method":"POST"},"RequestUri":"http://127.0.0.1:8001/api/v1/namespaces/my-namespace/serviceaccounts","Properties":{},"Content":"{\n  \"apiVersion\": \"v1\",\n  \"kind\": \"ServiceAccount\",\n  \"metadata\": {\n    \"labels\": {\n      \"mindaro.io/version\": \"v2\",\n      \"mindaro.io/component\": \"lpkrestorationjob\"\n    },\n    \"name\": \"lpkrestorationjob-v2\",\n    \"namespace\": \"my-namespace\"\n  }\n}","Headers":{"Content-Type":["application/json; charset=utf-8"],"Content-Length":["246"]}},"Response":{"StatusCode":400,"ReasonPhrase":"Bad Request","Content":"{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"the object provided is unrecognized (must be of type ServiceAccount): couldn't get version/kind; json parse error: unexpected end of JSON input (5b0a20207b0a202020202276616c7565223a205b5d2c0a2020202022706 ...)\",\"reason\":\"BadRequest\",\"code\":400}\n","Headers":{"Server":["SimpleHTTP/0.6","Python/3.6.9"],"Date":["Tue, 24 Nov 2020 14:25:40 GMT","Tue, 24 Nov 2020 14:25:40 GMT"],"Audit-Id":["fc19d11d-a5cd-4092-b7eb-502e3fffb8b0"],"Cache-Control":["no-cache, private"],"Content-Length":["325"],"Content-Type":["application/json"]}},"Body":null,"StackTrace":"   at k8s.Kubernetes.CreateNamespacedServiceAccountWithHttpMessagesAsync(V1ServiceAccount body, String namespaceParameter, String dryRun, String fieldManager, String pretty, Dictionary`2 customHeaders, CancellationToken cancellationToken)\n   at k8s.KubernetesExtensions.CreateNamespacedServiceAccountAsync(IKubernetes operations, V1ServiceAccount body, String namespaceParameter, String dryRun, String fieldManager, String pretty, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Common.Kubernetes.KubernetesClient.<>c__DisplayClass37_0.<<CreateServiceAccountIfNotExists>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.DevSpaces.Common.Kubernetes.KubernetesClient.<>c__DisplayClass47_0`1.<<ClientInvokeWrapperAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.DevSpaces.Common.Utilities.WebUtilities.RetryUntilTimeAsync(Func`2 action, TimeSpan maxWaitTime, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Common.Kubernetes.KubernetesClient.ClientInvokeWrapperAsync[T](Func`1 handler, String operation, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Library.Connect.RemoteRestoreJobDeployer._EnsureRbacResourcesAsync(String namespaceName, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Library.Connect.RemoteRestoreJobDeployer.CreateRemoteRestoreJobAsync[T](String targetName, String namespaceName, T patch, CancellationToken cancellationToken)\n   at Microsoft.DevSpaces.Library.Connect.KubernetesRemoteEnvironmentManager._PatchDeploymentAsync(RemoteContainerConnectionDetails remoteContainerConnectionDetails, IProgress`1 progress, CancellationToken cancellationToken)","Message":"Operation returned an invalid status code 'BadRequest'","Data":{},"InnerException":null,"HelpLink":null,"Source":"KubernetesClient","HResult":-2146233088},"HelpLink":null,"Source":null,"HResult":-2146233088}
2020-11-24T14:25:41.3478937Z | Library | TRACE | Event: KubernetesRemoteEnvironmentManager-Restore <json>{"properties":{"result":"Succeeded"},"metrics":{"durationInMs":0.0}}</json>
2020-11-24T14:25:41.3480607Z | Library | TRACE | Event: KubernetesRemoteEnvironmentManager-RestoreRemoteAgent <json>{"properties":{"result":"Succeeded"},"metrics":{"durationInMs":6.0}}</json>
2020-11-24T14:25:41.3481507Z | Library | TRACE | Event: ConnectManagementClient-RestoreOriginalRemoteContainer <json>{"properties":{"result":"Succeeded"},"metrics":{"durationInMs":7.0}}</json>
...
```
When I execute the request directly (via curl), it gives me a 409 (serviceaccounts "lpkrestorationjob-v2" already exists).

Is this a known issue, or related to this issue or new issue (for which I should create a new issue)?

Thanks for your help

**Comment by daniv-msft** [comment]
@creydr Thanks for your reply, and sorry you're encountering this issue. Unfortunately as of today we don't support the `kubectl proxy` scenario as a way to bypass the C# KubernetesClient authentication issues.

This is one of the options we will consider when looking for workarounds, but as of today I don't see how it would work without us explicitly changing code on our side to make it happen.

**Comment by creydr** [comment]
Hi @daniv-msft,
ok. Thanks for your response.
When I try to use a context with a kubernetes-admin user (without configured OIDC - only for testing), it does not start as well:

The *-restore pod prints in a loop the following:
```
2020-11-24T18:05:47.1162899Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"GetV1DeploymentAsync","success":true,"duration":null,"properties":{}}</json>
2020-11-24T18:05:47.1193557Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"ListPodsInNamespaceAsync","success":true,"duration":null,"properties":{}}</json>
2020-11-24T18:05:47.1194743Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"ListPodsForDeploymentAsync","success":true,"duration":null,"properties":{}}</json>
2020-11-24T18:05:47.1205248Z | RestorationJob | TRACE | Event: RestorationJob-AgentPing <json>{"eventName":"RestorationJob-AgentPing","properties":{"restorePerformed":"false","numFailedPings":"0","hasConnectedClients":"false","result":"Succeeded"},"metrics":{"durationInMs":9.0}}</json>
```

while the log in bridge-library-x.txt says (in a loop) the following:
```
2020-11-24T18:08:57.7912034Z | Library | TRACE | Accept 37983 to 50051
2020-11-24T18:08:57.7928770Z | Library | TRACE | Port forward ticket-service-69d56db5bf-tmmwz 37983:50051 59518 : Run send loop
2020-11-24T18:08:57.7944414Z | Library | TRACE | Creating web socket for ticket-service-69d56db5bf-tmmwz 50051
2020-11-24T18:08:58.3887471Z | Library | TRACE | Dependency: Kubernetes <json>{"target":"WebSocketPodPortForwardAsync","success":true,"duration":null,"properties":{}}</json>
2020-11-24T18:08:58.3890579Z | Library | TRACE | Web socket for ticket-service-69d56db5bf-tmmwz 50051 created.
2020-11-24T18:08:58.3902005Z | Library | TRACE | Port forward ticket-service-69d56db5bf-tmmwz 37983:50051 59518 : Run receive loop
2020-11-24T18:09:16.9985232Z | Library | TRACE | Port forward ticket-service-69d56db5bf-tmmwz 37983:50051 58582 : RemoteConnection closed.
2020-11-24T18:09:16.9992571Z | Library | TRACE | Port forward ticket-service-69d56db5bf-tmmwz 37983:50051 58582 : Remote stream finished. Closed
2020-11-24T18:09:20.0978192Z | Library | TRACE | Accept 37983 to 50051
....
```

Can you help me with this issue here, or should I create a new issue for this?

Thanks

**Comment by jarrett-confrey** [comment]
@daniv-msft @greenie-msft it appears that there is now oidc support in the kubernetes csharp client with https://github.com/kubernetes-client/csharp/pull/544 🎉 i would love to see this pulled in so that i can start playing with bridge to kubernetes in vscode on my mac 🥺

**Comment by arbielsk** [comment]
@jarrett-confrey afaik this is already available and works (last test about a month ago).

I was not able to verify that routing-manager works as our cluster-admin-team needs to make adjustments to our k8s cluster first.

**Comment by jarrett-confrey** [comment]
thank you @arbielsk! i assumed i was seeing the same issue. i have opened https://github.com/microsoft/mindaro/issues/174 for the issue i am seeing

**Comment by jarrett-confrey** [comment]
i was able to fix the issue with https://github.com/kubernetes-client/csharp/pull/633
